Iteration 5

andreevym opened the pull request.

andreevym committed:
#1 - added go.mod
- added http server
- added mem storage
- added api post /update for store data for metric types gauge and counter

andreevymn-t-io committed:
# 2 added client

andreevymn-t-io committed:
#3 added get value method "/value/{metricType}/{metricName}", change server implementation from "net/http" to "chi"

andreevymn-t-io committed:
#4 flag add for server and for client

andreevymn-t-io committed:
#4 fix linter - close body

andreevymn-t-io committed:
# 4 fix linter - don't use unit-specific suffix "Sec"

andreevymn-t-io committed:
# 4 fix linter - move defer

andreevymn-t-io committed:
# 5 handle env for agent

andreevymn-t-io committed:
# 5 add module github.com/caarlos0/env

andreevymn-t-io committed:
# 5 change handle send metric to server

andreevymn-t-io committed:
# 5 fix vet linter

andreevymn-t-io committed:
# 5 update README.md

andreevymn-t-io committed:
# 5 update go tests

andreevymn-t-io committed:
# 5 update go tests

andreevymn-t-io committed:
# 5 fix linter body close

andreevymn-t-io committed:
# 5 fix problem in /value/counter/test , /value/gauge/test with return array in result

andreevymn-t-io committed:
# 3 add test for get /value/ for unknown metricName and metricType

andreevymn-t-io committed:
# 3 update storage

andreevym committed:
# 3 add tests for gauge and counter

andreevymn-t-io committed:
# 3 fix counter, need to sum old and new val

andreevymn-t-io committed:
# 3 fix tests counter, need to sum old and new val

andreevymn-t-io committed:
# 4 updated flag type

andreevymn-t-io committed:
# 4 updated move protocol to request

andreevymn-t-io committed:
# 4 add address env for server

andreevymn-t-io committed:
# 4 fix README.md

andreevymn-t-io committed:
# 4 add section "Contents"

capsulated commented on cmd/agent/env.go:
```diff
@@ -0,0 +1,7 @@
+package main
+
+type EnvConfig struct {
```
Конфиги тоже можно в отдельный пакет переместить

andreevym replied:
исправил

capsulated commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,94 @@
+	"github.com/caarlos0/env"
+)
+
+var (
```
❌ Эти переменные и константы необходимо перенести в соответсвующий пакет. `cmd` - это инициализация и запуск приложения и каких-то особенностей реализации в  `cmd` быть не должно

andreevym replied:
исправил, перенес в отдельный пакет.

capsulated commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,94 @@
+
+const (
+	// ContentType Заголовок — Content-Type: text/plain.
+	ContentType = "text/plain"
```
❌ Это явно в хендлеры перенести нужно

andreevym replied:
исправил, перенес в хендлеры

capsulated commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,94 @@
+
+func main() {
+	// парсим переданные серверу агенту в зарегистрированные переменные
+	flag.Parse()
```
Вот хорошо бы было весь парсинг и проверка конфигов была в отдельном пакете и методе, а здесь только вызов этого метода

andreevym replied:
постарался вынести в отдельный пакет config, правда получилось не очень красиво как по мне

capsulated replied:
init() перед main() не особо красиво, а в остальном вроде всё ок

capsulated commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,94 @@
+	time.Sleep(time.Minute)
+}
+
+func sendMetricToServer(tickerReport *time.Ticker, metricServerURL string) {
```
⚠️ Очень желательно создать отдельный пакет для агента, так называемый воркер. Описывать его в main() пакета `cmd` плохая идея

andreevym replied:
сделал отдельный пакет metric и туда перенес функции `sendByTickerAndAddress` и `pollLastMemStatByTicker`

capsulated commented on internal/repository/repository.go:
```diff
@@ -0,0 +1,8 @@
+package repository
+
+type Storage interface {
```
⚠️ Интерфейсы в go [обычно объявляются в пакете потребителя интерфейса](https://github.com/golang/go/wiki/CodeReviewComments#interfaces)

capsulated commented on internal/handlers/server.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/andreevym/metric-collector/internal/repository"
+)
+
+type Server struct {
```
❌ Получается путаница с объектами - два сервера в приложении, в разных пакетах. Скорее всего текущий нужно переименовать.

Кажется, лишний объект. Его логика, скорее всего, должна быть скрыта в реализации хранилки. Или в отдельном пакете для метрик. Но точно не в хендлерах. Лучше полностью избавиться от этой структуры

andreevym replied:
переименован в ServiceHandlers

andreevym replied:
переименован в ServiceHandlers

andreevym replied:
переименован в ServiceHandlers

capsulated commented on internal/handlers/router.go:
```diff
@@ -0,0 +1,16 @@
+
+import "github.com/go-chi/chi"
+
+func Router(s Server) chi.Router {
```
⚠️ Очень важно разобраться, [когда используются указатели в ресивере](https://go.dev/doc/faq#methods_on_values_or_pointers), и когда указатели в выходных значениях. Чаще всего, это указатели, если это не простые типы

andreevym replied:
разобрался

capsulated commented on internal/handlers/router.go:
```diff
@@ -0,0 +1,16 @@
+	r := chi.NewRouter()
+	r.Handle(
+		"/update/{metricType}/{metricName}/{metricValue}",
+		s.UpdateMetricHandler(),
```
Соответсвенно, если убрать из хендлера возвращение функции, то тут, скорее всего будет  так:
```go
r.Handle("/update/{metricType}/{metricName}/{metricValue}", s.UpdateMetricHandler)
```

andreevym replied:
испавил. да, всё верно.

capsulated commented on internal/handlers/update_handler.go:
```diff
@@ -0,0 +1,57 @@
+// UpdateMetricHandler method for insert or update metrics
+// example request url: http://<АДРЕС_СЕРВЕРА>/update/<ТИП_МЕТРИКИ>/<ИМЯ_МЕТРИКИ>/<ЗНАЧЕНИЕ_МЕТРИКИ>
+func (s Server) UpdateMetricHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
```
Не совсем понятно, для чего возвращается функция (есть комментарий про это в роутере)

andreevym replied:
исправил

capsulated commented on internal/model/metric.go:
```diff
@@ -0,0 +1,6 @@
+package model
+
+const (
```
В го не принято объявлять модели отдельно от их использования. Обычно, также как с интерфейсами: там где есть потребитель модели, там и объявляется

andreevym replied:
исправил, теперь пакета model нет

capsulated reviewed (changes requested):
❌ - комментарии, которые обязательны к исправлению
⚠️ - комментарии, с которыми очень желательно разобраться и исправить

Плюс, обсудим эти моменты на 1-1

andreevymn-t-io committed:
# 5 fix review comment
- removed model pkg
- renamed Server to ServiceHandlers
- moved config to separate dir
- made multistorage struc to handle different type of storage or store type
- create metric pkg
- move to metric pkg UpdateMetricContentType
- fix problem with http.HandlerFunc

andreevymn-t-io committed:
# 5 fix linter

andreevymn-t-io committed:
# 5 fix linter and env setup

andreevymn-t-io committed:
# 5 fix metric agent

capsulated commented on cmd/server/main.go:
```diff
@@ -1,3 +1,27 @@
+)
+
+func init() {
+	config.ServerFlags()
```
да можно и в main() перенести как мне кажется

andreevym replied:
готово

capsulated commented on internal/multistorage/save.go:
```diff
@@ -0,0 +1 @@
+package multistorage
```
Какой-то странный пустой файл 😳

andreevym replied:
избавились от этого файла!

capsulated reviewed (approved):
Все комменты исправлены 🔥 Оставил пару новых комментариев, там совсем не критично. 

В следующем спринте будем смотреть , удобно ли получилось))

andreevymn-t-io committed:
# 5 fix metric agent

andreevymn-t-io committed:
# 5 remove init method from cmd
